{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deliverables:  36 total points\n",
    "\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as 4(or 5) .py files, \n",
    "     \n",
    " Due: Monday January 20, 2020 11:55pm\n",
    "\n",
    "# Lab 2 – Manipulating Data Types (36 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment, and as a final step we will create distinct command line programs from each of the parts.\n",
    "Please save each of the four (or five) python programs as .py files into your LAB02 folder. You can copy the text from the code cell, then paste it into an editor like notepad or textedit. Save that new file with the appropriate name ( eg. seqCleaner.py) into your LAB02 folder. Using terminal or cmd or conda, you can now do a final test on that new program with the command:<br>\n",
    "python3 seqCleaner.py <br>\n",
    "\n",
    "All together, you will submit:\n",
    " - seqCleaner.py, \n",
    " - fastqParse,py, \n",
    " - coordinateMathSoln.py, \n",
    " - converter.py, and \n",
    " - (optionally) Triad.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence cleanup\n",
    "In this exercise, you will create a program to “clean up” a sequence of DNA by removing ambiguous bases (denoted by “N”) output from a sequencer. Your task is to create a Python program called seqCleaner that\n",
    " - asks for and collects a sequence of DNA using input()\n",
    " - removes the ambiguous parts of the sequence, outputs the “cleaned” sequence, replacing the ambiguous parts with a count in {}’s.\n",
    "For example, if I enter the sequence of DNA “AaNNNNNNGTC” (without quotes), the program will output:<br>\n",
    "AA{6}GTC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hints: \n",
    " - The input sequence is not guaranteed to be uppercase, but should be interpreted as though it is all uppercase.\n",
    " - Only the letters (A,C,G,T,N) will be included in the input.\n",
    " - Only the DNA characters (A,C,G,T) should remain after the cleanup. \n",
    " - The input may include, at most, one block of 'N' characters. \n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and specific line or block comments that explain the semantics of your implementation.\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs. This should be in the form of a  program level dosctring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA data? atcnnnngta\n",
      "ATC{4}GTA\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Marissa Glover (mlglover) \n",
    "# Group Members: Parissa Monem (pmonem) and Matt Modena (mmodena)\n",
    "\n",
    "'''\n",
    "Read a DNA string from user input and return a collapsed substring of embedded Ns to: {count}.\n",
    "\n",
    "Example: \n",
    " input: AaNNNNNNGTC\n",
    "output: AA{6}GTC\n",
    "\n",
    "Any lower case letters are converted to uppercase.\n",
    "'''\n",
    "\n",
    "class DNAstring (str):\n",
    "    '''Determine the number of Ns in a string and create a new object.\n",
    "    A single run of Ns will be collapsed and show {#Ns}.'''\n",
    "    def purify(self):\n",
    "        '''Given an upper case string, this will find the first and last N.\n",
    "        The string will be split at the Ns and then merged back together with {#Ns} in place of the Ns.'''\n",
    "        numN = self.count('N')\n",
    "        numNstring = str({numN})\n",
    "        findFirstN = self.find('N')\n",
    "        findLastN = self.rfind('N')\n",
    "        splitSeq = self.split('N')\n",
    "        beforeN = splitSeq[0]\n",
    "        afterN = splitSeq[-1]\n",
    "        pureDNA = ''.join([beforeN, numNstring, afterN])\n",
    "        return pureDNA\n",
    "    \n",
    "def main():\n",
    "    '''Get user DNA data and clean it up. Make it upper case before passing to the class.'''\n",
    "    data = input('DNA data? ')\n",
    "    upperDNA = data.upper()\n",
    "    thisDNA = DNAstring (upperDNA)\n",
    "    pureData = thisDNA.purify()\n",
    "    print (pureData)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence information parsing\n",
    "In this exercise, you will create a program to “parse” sequence name information from a single line of a FASTQ formatted file. Your task is to create a Python script called fastqParse that:\n",
    " - asks for and collects the seqname line of a FASTQ file using input()\n",
    " - parses out each field of the run information from the string and displays each of them on a new line\n",
    "For example, if I enter the FASTQ seqname line:<br> \n",
    "__*@EAS139:136:FC706VJ:2:2104:15343:197393*__\n",
    "then the program will output:<br>\n",
    "Instrument = EAS139<br>\n",
    "Run ID = 136<br>\n",
    "Flow Cell ID = FC706VJ<br>\n",
    "Flow Cell Lane = 2<br>\n",
    "Tile Number = 2104<br>\n",
    "X-coord = 15343<br>\n",
    "Y-coord = 197393\n",
    "### Hints: \n",
    " - The input string is guaranteed to have 7 fields. \n",
    " - The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”.  \n",
    " - A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain documentation/comments\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fastq seqname line? @EAS139:136:FC706VJ:2:2104:15343:197393\n",
      "Instrument = EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell Lane = 2\n",
      "Tile Number = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Your full name (CATS account username) \n",
    "# Group Members: Parissa Monem (pmonem) and Matt Modena (mmodena)\n",
    "\n",
    "'''\n",
    "Read a DNA string from user input and return a collapsed substring of embedded Ns to: {count}.\n",
    "\n",
    "Example: \n",
    " input: SeqName line from fastq file \n",
    "     (example: @EAS139:136:FC706VJ:2:2104:15343:197393)\n",
    "output: each field of the run information parsed out displaying each of them on a new line\n",
    "    (example:   Instrument = EAS139\n",
    "                Run ID = 136\n",
    "                Flow Cell ID = FC706VJ\n",
    "                Flow Cell Lane = 2\n",
    "                Tile Number = 2104\n",
    "                X-coord = 15343\n",
    "                Y-coord = 197393)\n",
    "'''\n",
    "\n",
    "class FastqString (str):\n",
    "    '''Take a seqName line from a fastq file and parse the line.'''\n",
    "    def parse(self):\n",
    "        '''Parse a seqName line and output the run information onto different lines.'''\n",
    "        splitLine = self.strip('@').split(':')\n",
    "        print ('Instrument = {0}'.format(splitLine[0]))\n",
    "        print ('Run ID = {0}'.format(splitLine[1]))\n",
    "        print ('Flow Cell ID = {0}'.format(splitLine[2]))\n",
    "        print ('Flow Cell Lane = {0}'.format(splitLine[3]))\n",
    "        print ('Tile Number = {0}'.format(splitLine[4]))\n",
    "        print ('X-coord = {0}'.format(splitLine[5]))\n",
    "        print ('Y-coord = {0}'.format(splitLine[6]))\n",
    "        \n",
    "def main():\n",
    "    '''Get input seqName line and return the parsed output.'''\n",
    "    fastqSeqName = input('Fastq seqname line? ')\n",
    "    fastqString = FastqString (fastqSeqName)\n",
    "    parsed = fastqString.parse()\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein coordinates\n",
    "In this exercise, you will create a program that takes three sets of atomic coordinates, all provided on a **single line**. The program then calculates the bond lengths and angles. For this program, you can start with the Triad class (provided). Your task is to create a Python program called coordinateMathSoln that:\n",
    " - asks for and collects three sets of coordinates using input(), only use 1 line for this data !!\n",
    " - outputs the N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits (see below)\n",
    " \n",
    "For example, if I enter the following coordinates (notice.. they are all on one line !!!) :<br>\n",
    "\n",
    "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465) <br>\n",
    "\n",
    "then the program will output the following three lines:<br>\n",
    "N-C bond length = 1.33<br>\n",
    "N-Ca bond length = 1.46<br>\n",
    "C-N-Ca bond angle = 124.0<br> \n",
    "(Note: make sure that the angle returned is in **degrees** !!)\n",
    "### Hints: \n",
    " - Each coordinate will contain only 3 numeric values. \n",
    " - Bond lengths are the distance between two points, which for points P and Q in 3-space,\n",
    "$(P_x , P_y , P_z )$ and $(Q_x , Q_y , Q_z )$ respectively, the distance between them is:\n",
    "\n",
    "\n",
    "$$\n",
    "\\begin {eqnarray}\n",
    "     \\|PQ\\| &=&\\sqrt{(P_x-Q_x)^2 + (P_y-Q_y)^2 + (P_z-Q_z)^2)} \\\\\n",
    "            &=&\\sqrt{\\sum^{i \\in x,y,z}{(P_i-Q_i)^2}} \\\\\n",
    "\\end{eqnarray}\n",
    "$$\n",
    "\n",
    " - Bond angles can be calculated from the dot product.\n",
    " \n",
    "Let’s say that we have three points in space labeled P, Q and R. We are interested in the angle at point Q that is made from line segments QP and QR.\n",
    "The dot product tells us (for standard vectors P and R) that:\n",
    "$$P \\cdot R=\\|P\\| \\|R\\| \\cos{\\theta}$$\n",
    "in this notation,$\\|P\\|$ refers to the length of vector P as a standard vector (assumed to begin at the origin (0,0,0) ).\n",
    "We can then see that the angle between vectors P and R can be found by:\n",
    "$$\\cos{\\theta}=\\frac{P \\cdot R} {\\|P\\| \\|R\\|}$$\n",
    "\n",
    "We can calculate the dot product using the sum of products of the vector components:\n",
    "$$P \\cdot R = \\sum^{i \\in x,y,z}{P_i R_i} $$\n",
    "\n",
    "Now, to find vector P in standard form, we need to remember that QP starts at Q, so we need to place the origin at Q and find out where P is in that new space. We do that by subtracting the components of Q from P.\n",
    "Putting all of this together, we get:\n",
    "$$\\theta=\\cos^{-1}{\\frac{\\sum^{i \\in x,y,z}{(P_i - Q_i) (R_i - Q_i) }} {\\|QP\\| \\|QR\\|}}$$\n",
    "Remember, $\\theta$ is in radians.\n",
    "\n",
    "Below I have given you a class (Triad) with methods to calculate dot products (dot), dot products of translated vectors (ndot), distances (dPQ, dPR, dPR) and angles in radians (angleP, angleQ and angleR) for each of the three points in a Triad object.\n",
    "A reasonable solution for this exercise involves around 12 lines of additional code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Three sets of coordinates: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-Ca bond length = 2.47\n",
      "C-N-Ca bond angle = 124.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Marissa Glover (mlglover)\n",
    "# Group Members: Parissa Monem (pmonem) and Matt Modena (mmodena)\n",
    "\n",
    "'''\n",
    "Calculate distances between points and angles given a triad of protein coordinates.\n",
    "\n",
    "input: x,y,z coordinates for a carbon, nitrogen, and alpha carbon of a protein. \n",
    "    (example: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465))\n",
    "output: N-C bond length, N-Ca bond length, and C-N-Ca bond angle in degrees.\n",
    "    (example:   N-C bond length = 1.33\n",
    "                N-Ca bond length = 1.46\n",
    "                C-N-Ca bond angle = 124.0)\n",
    "'''\n",
    "\n",
    "import math\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) /   math.sqrt(self.d2(self.q,self.p)*self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /  math.sqrt(self.d2(self.p,self.q)*self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /  math.sqrt(self.d2(self.p,self.r)*self.d2(self.q,self.r)))\n",
    "\n",
    "def main():\n",
    "    '''Ask for an input with three sets of coordinates and parse the string.\n",
    "    Replace parentheses and equals signs with commas, and replace spaces with nothing.\n",
    "    Then we can split the line at the commas to get a list of strings.'''\n",
    "    dataInput = input('Three sets of coordinates: ')\n",
    "    splitLine = dataInput.replace('(',',').replace(')',',').replace('=',',').replace(' ','').split(',')\n",
    "    #Use map function to turn list of strings into list of floats\n",
    "    #https://www.thecodingforums.com/threads/how-to-convert-a-list-of-strings-to-a-tuple-of-floats.684531/\n",
    "    p = tuple(map(float, splitLine[2:5]))\n",
    "    q = tuple(map(float, splitLine[7:10]))\n",
    "    r = tuple(map(float, splitLine[12:15]))\n",
    "    triad = Triad(p,q,r)\n",
    "    print ('N-C bond length = {0:0.2f}'.format(triad.dPQ()))\n",
    "    print ('N-Ca bond length = {0:0.2f}'.format(triad.dQR()))\n",
    "    #Convert C-N-Ca bond angle from radians to degrees with math\n",
    "    print ('C-N-Ca bond angle = {0:0.1f}'.format(math.degrees(triad.angleQ())))\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra credit (5 points): Rewrite the Triad class.  \n",
    "For extra-credit, provide a direct replacement for the Triad class. The external methods that calculate angles, distances, and points (tuples) p,q and r must be maintained such that either version of the Triad class can be used.\n",
    "\n",
    "You could use the cosine law to calculate angles instead of the dot product.  You might make use of the numpy module. You might recode each of the methods to avoid using zip. You might consider using list iterations.\n",
    "Your Triad replacement must reimplement all of Triad public function, without using zip and without being a trivial rewrite. Your implementation need not be as compact as the current implementation, and it needs to be correct and fully documented to receive full credit. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Triad rewrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Three sets of coordinates: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-N-Ca bond angle = 124.0\n",
      "bond angle P = 29.4\n",
      "bond angle R = 26.6\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Marissa Glover (mlglover)\n",
    "# Group Members: None\n",
    "'''\n",
    "Create a different version of triad that can be used as an alternative to the previous Triad class.\n",
    "\n",
    "\n",
    "input: x,y,z coordinates for a carbon, nitrogen, and alpha carbon of a protein. \n",
    "    (example: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465))\n",
    "output: N-C bond length, N-Ca bond length, and C-N-Ca bond angle in degrees.\n",
    "    (example:   N-C bond length = 1.33\n",
    "                N-Ca bond length = 1.46\n",
    "                C-N-Ca bond angle = 124.0)\n",
    "'''\n",
    "\n",
    "#import sys\n",
    "#!conda install --yes --prefix {sys.prefix} numpy\n",
    "import numpy as np\n",
    "import math\n",
    "class Triad:\n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad.\n",
    "\n",
    "        Convert the tuple for coordinates of each atom into array.\n",
    "        Needs to use asarray function, not array (array will not convert input will only create new array).\n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ).\n",
    "        \"\"\"\n",
    "        self.p = np.asarray(p)\n",
    "        self.q = np.asarray(q)\n",
    "        self.r = np.asarray(r)\n",
    "        \n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q.\n",
    "        Find the difference between p and q points in space which have been converted to arrays (2D to 3D).\n",
    "        \n",
    "        Use the linear algebra functions from NumPy to use standard linear algebra algorithms.\n",
    "        (from https://stackoverflow.com/questions/1401712/how-can-the-euclidean-distance-be-calculated-with-numpy)\n",
    "        \n",
    "        A norm is a function that satisfies certain properties pertaining to scalability and additivity,\n",
    "        and assigns a strictly positive real number to each vector in a vector space over the field of real or complex numbers\n",
    "        —except for the zero vector, which is assigned zero (from https://en.wikipedia.org/wiki/Norm_(mathematics)).\n",
    "        Analogous to absolute value.\n",
    "        linalg.norm(x[, ord, axis, keepdims]) for matrix or vector norm.\n",
    "        \"\"\"\n",
    "        return np.linalg.norm(self.p - self.q)\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r.\n",
    "        The same as the dPQ function, but for points p and r.\n",
    "        \"\"\"\n",
    "        return np.linalg.norm(self.p - self.r)\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r.\n",
    "        The same as the dPQ function, but for points q and r.\n",
    "        \"\"\"\n",
    "        return np.linalg.norm(self.q - self.r)\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians).\n",
    "        \n",
    "        cosAngle = (PQ.PR)/(|PQ||PR|)\n",
    "        Use dot function to get dot product between QP and QR (input as arrays).\n",
    "            This must be achieved by doing dot products of point q - point p and point r - point p.\n",
    "            The point for the angle we are looking for needs to come last in subtraction.\n",
    "            point q - point p is not the same as point p - point q.\n",
    "            (Subtraction of vectors from https://www.python-course.eu/matrix_arithmetic.php)\n",
    "        Then divide by the product of PQ and PR.\n",
    "        To solve for the angle, the function must determine the inverse cosine.\n",
    "        \"\"\"\n",
    "        distPQ = np.linalg.norm(self.p - self.q)\n",
    "        distPR = np.linalg.norm(self.p - self.r)\n",
    "        q_minus_p = self.q - self.p\n",
    "        r_minus_p = self.r - self.p\n",
    "        return math.acos(np.dot(q_minus_p, r_minus_p) / (distPQ * distPR))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians).\n",
    "        The same as the dPQ function, but for points p and r.\n",
    "        \"\"\"\n",
    "        distPQ = np.linalg.norm(self.p - self.q)\n",
    "        distQR = np.linalg.norm(self.q - self.r)\n",
    "        p_minus_q = self.p - self. q\n",
    "        r_minus_q = self.r - self.q\n",
    "        return math.acos(np.dot(p_minus_q, r_minus_q) / (distPQ * distQR))\n",
    "        \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians).\n",
    "        The same as the dPQ function, but for points p and r.\"\"\"\n",
    "        distPR = np.linalg.norm(self.p - self.r)\n",
    "        distQR = np.linalg.norm(self.q - self.r)\n",
    "        p_minus_r = self.p - self.r\n",
    "        q_minus_r = self.q - self.r\n",
    "        return math.acos(np.dot(p_minus_r, q_minus_r) / (distPR * distQR))\n",
    "\n",
    "def main():\n",
    "    '''Ask for an input with three sets of coordinates and parse the string.\n",
    "    Replace parentheses and equals signs with commas, and replace spaces with nothing.\n",
    "    Then we can split the line at the commas to get a list of strings.'''\n",
    "    dataInput = input('Three sets of coordinates: ')\n",
    "    splitLine = dataInput.replace('(',',').replace(')',',').replace('=',',').replace(' ','').split(',')\n",
    "    #Use map function to turn list of strings into list of floats\n",
    "    #https://www.thecodingforums.com/threads/how-to-convert-a-list-of-strings-to-a-tuple-of-floats.684531/\n",
    "    p = tuple(map(float, splitLine[2:5]))\n",
    "    q = tuple(map(float, splitLine[7:10]))\n",
    "    r = tuple(map(float, splitLine[12:15]))\n",
    "    triad = Triad(p,q,r)\n",
    "    print ('N-C bond length = {0:0.2f}'.format(triad.dPQ()))\n",
    "    print ('N-Ca bond length = {0:0.2f}'.format(triad.dQR()))\n",
    "    #Convert C-N-Ca bond angle from radians to degrees with math\n",
    "    print ('C-N-Ca bond angle = {0:0.1f}'.format(math.degrees(triad.angleQ())))\n",
    "\n",
    "main()     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon tables and amino acid letter converters\n",
    "In this exercise, you will create a program that uses mappings to convert sequence information between different amino acid representations.  This includes the 3-letter codon code (RNA and DNA), the one letter amino acid code and the 3-letter amino acid code.\n",
    "The program will use different dictionaries that represent :\n",
    "codon tables – one for DNA and the other for RNA, and\n",
    "amino acid letter representation converters. \n",
    "\n",
    "Your task is to create a Python program called converter that\n",
    "asks for and collects a single input string using input()\n",
    "parses the string, looks up the information in the appropriate dictionary, and outputs the correct conversion\n",
    "For example:<br>\n",
    "\n",
    "if I enter “ATG” (without quotes), then the program will output:<br>\n",
    "ATG = MET<br>\n",
    " \n",
    "if I enter “UAG” (without quotes), then the program will output:<br>\n",
    "UAG = ---<br>\n",
    "if I enter “E” (without quotes), then the program will output:<br>\n",
    "E = GLU<br>\n",
    "if I enter “Asp” (without quotes), then the program will output:<br>\n",
    "ASP = D<br>\n",
    "### Hints: \n",
    " - The program might not get a valid codon.  In that case, it should output 'unknown'.  You can use the dictionary method 'get' and include a default_value to handle the 'unknown' case.  See Model p. 72 for an example.\n",
    " \n",
    "To get full credit on this assignment, your program needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA codon, RNA codon, or amino acid = x\n",
      "X = unknown\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Marissa Glover (mlglover)\n",
    "# Group Members: Parissa Monem (pmonem) and Matt Modena (mmodena)\n",
    "\n",
    "'''\n",
    "Convert sequence information between different amino acid representations.\n",
    "This will parse the input and search dictionaries of amino acids and codons for the correct conversion.\n",
    "The program will not return codons for amino acid inputs though.\n",
    "The program also will only recognize the 20 common amino acids (and not things like selenocysteine).\n",
    "If an invalid input is given, the program will return unknown.\n",
    "\n",
    "input: single input string of 3 letter codon, 3 letter amino acid, or 1 letter amino acid\n",
    "output: correct conversion to 1 letter amino acid or 3 letter amino acid\n",
    "'''\n",
    "short_AA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }\n",
    "\n",
    "long_AA = {value:key for key,value in short_AA.items()}\n",
    "\n",
    "rnaCodonTable = {\n",
    "# Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "def main():\n",
    "    '''Ask for an input codon or amino acid and parse the string.\n",
    "    Search each dicitonary for the presence of the input.\n",
    "    If the input is invalid, return unknown.'''\n",
    "    data = input ('DNA codon, RNA codon, or amino acid = ')\n",
    "    upperData = data.upper()\n",
    "    findValue = short_AA.get(upperData) or long_AA.get(upperData) or rnaCodonTable.get(upperData) or dnaCodonTable.get(upperData) or dnaCodonTable.setdefault(upperData,'unknown')\n",
    "    print (upperData + ' = ' + findValue)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The two recommended solutions (pick one) at this point are to:\n",
    " - Email your code files to yourself\n",
    " - Copy your code files to a pen drive\n",
    " - Save your work on your own computer if you are using Anaconda ( or equiv)\n",
    "\n",
    "The class will be using canvas to submit the source code files created as part of this lab assignment.\n",
    "For this lab, you should upload the following programs as files:\n",
    " \n",
    " - seqCleaner.py 9 points\n",
    " - fastqParse.py 9 points\n",
    " - coordinateMathSoln.py 9 points\n",
    " - converter.py 9 points\n",
    " - triad.py Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Three sets of coordinates: felton = (37.0513, -122.0733) santacruz = (36.984532, -121.963123) campbell = (37.2871651, -121.9499568)\n",
      "(37.0513, -122.0733) (36.984532, -121.963123) (37.2871651, -121.9499568)\n",
      "felton to santa cruz = 0.13\n",
      "felton to campbell = 0.27\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) /   math.sqrt(self.d2(self.q,self.p)*self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /  math.sqrt(self.d2(self.p,self.q)*self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /  math.sqrt(self.d2(self.p,self.r)*self.d2(self.q,self.r)))\n",
    "\n",
    "def main():\n",
    "    '''Ask for an input with three sets of coordinates and parse the string.\n",
    "    Replace parentheses and equals signs with commas, and replace spaces with nothing.\n",
    "    Then we can split the line at the commas to get a list of strings.'''\n",
    "    dataInput = input('Three sets of coordinates: ')\n",
    "    splitLine = dataInput.replace('(',',').replace(')',',').replace('=',',').replace(' ','').split(',')\n",
    "    p = tuple(map(float, splitLine[2:4]))\n",
    "    q = tuple(map(float, splitLine[6:8]))\n",
    "    r = tuple(map(float, splitLine[10:12]))\n",
    "    print (p,q,r)\n",
    "    triad = Triad(p,q,r)\n",
    "    print ('felton to santa cruz = {0:0.2f}'.format(triad.dPQ()))\n",
    "    print ('felton to campbell = {0:0.2f}'.format(triad.dPR()))\n",
    "    #print ('C-N-Ca bond angle = {0:0.1f}'.format(math.degrees(triad.angleQ())))\n",
    "    #felton = (37.0513, -122.0733) santacruz = (36.984532, -121.963123) campbell = (37.2871651, -121.9499568)\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
